iterator-test-harness/src/main/java/org/apache/accumulo/iteratortest/testcases/ReSeekTestCase.java

```diff
@@ -0,0 +1,107 @@
+import java.util.Collections;
+import java.util.Random;
+import java.util.TreeMap;
+
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+import org.apache.accumulo.iteratortest.IteratorTestInput;
+import org.apache.accumulo.iteratortest.IteratorTestOutput;
+import org.apache.accumulo.iteratortest.IteratorTestUtil;
+import org.apache.accumulo.iteratortest.environments.SimpleIteratorEnvironment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Test case that verifies that an iterator can use the generated instance from {@code deepCopy}.
+ */
+public class ReSeekTestCase extends OutputVerifyingTestCase {
+  private static final Logger log = LoggerFactory.getLogger(ReSeekTestCase.class);
+
+  /**
+   * Let N be a random number between [0, RESEEK_INTERVAL). After every Nth entry "returned" to the client, recreate and reseek the iterator.
+   */
+  private static final int RESEEK_INTERVAL = 4;
+
+  private final Random random;
+
+  public ReSeekTestCase() {
+    this.random = new Random();
+  }
+
+  @Override
+  public IteratorTestOutput test(IteratorTestInput testInput) {
+    final SortedKeyValueIterator<Key,Value> skvi = IteratorTestUtil.instantiateIterator(testInput);
+    final SortedKeyValueIterator<Key,Value> source = IteratorTestUtil.createSource(testInput);
+
+    try {
+      skvi.init(source, testInput.getIteratorOptions(), new SimpleIteratorEnvironment());
+      skvi.seek(testInput.getRange(), Collections.<ByteSequence> emptySet(), false);
+      return new IteratorTestOutput(consume(skvi, testInput));
+    } catch (IOException e) {
+      return new IteratorTestOutput(e);
+    }
+  }
+
+  TreeMap<Key,Value> consume(SortedKeyValueIterator<Key,Value> skvi, IteratorTestInput testInput) throws IOException {
+    final TreeMap<Key,Value> data = new TreeMap<>();
+    final Range origRange = testInput.getRange();
+    int reseekCount = random.nextInt(RESEEK_INTERVAL);
+
+    int i = 0;
+    while (skvi.hasTop()) {
+      data.put(skvi.getTopKey(), skvi.getTopValue());
+
+      /*
+       * One of the trickiest cases in writing iterators:
+       * 
+       * After any result is returned from a TabletServer to the client, the Iterator in the TabletServer's memory may be torn down. To preserve the state and
+       * guarantee that all records are received, the TabletServer does remember the last Key it returned to the client. It will recreate the Iterator (stack),
+       * and seek it using an updated Range. This range's start key is set to the last Key returned, non-inclusive.
+       */
+      if (i % RESEEK_INTERVAL == reseekCount) {
+        // Last key
+        Key reSeekStartKey = skvi.getTopKey();
+
+        // Deepcopy the iterator
+        skvi = skvi.deepCopy(new SimpleIteratorEnvironment());
```

Comment:
creating a completely new iterator instance and initializing it would be more realistic